### PR TITLE
raftstore: make sure callback is called

### PR DESCRIPTION
**Pull request by BusyJay**
If callback is not called, it may lead to deadlock, which is unrecoverable, should panic in that case.

@siddontang @disksing @zhangjinpeng1987 @hhkbp2 PTAL

### CODE REVIEW

**Comment by BusyJay** [comment]
see also #1223

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -565,10 +576,16 @@ impl Peer {
 
             // transfer leader command doesn't need to replicate log and apply, so we
             // return immediately. Note that this command may fail, we can view it just as an advice
-            return cmd.cb.call_box((make_transfer_leader_response(),));
+            cmd.cb.take().unwrap().call_box((make_transfer_leader_response(),));
+            return false;
         } else if get_change_peer_cmd(&req).is_some() {
             if self.raft_group.raft.pending_conf {
-                return Err(box_err!("there is a pending conf change, try later"));
+                error!("{} there is a pending conf change, try later", self.tag);
```
Comment (ngaut) [inline src/raftstore/store/peer.rs]: s/error/info/

**Comment by ngaut** [comment]
Rest LGTM.
PTAL @zhangjinpeng1987 @siddontang

**Comment by zhangjinpeng1987** [comment]
LGTM

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -139,17 +147,15 @@ impl PendingCmdQueue {
...
     let mut resp = cmd_resp::new_error(region_not_found);
     cmd_resp::bind_uuid(&mut resp, cmd.uuid);
     debug!("[region {}] {} is removed, notify {}.",
            region_id,
            peer_id,
            cmd.uuid);
-    if let Err(e) = cmd.cb.call_box((resp,)) {
-        error!("failed to notify {}: {:?}", cmd.uuid, e);
-    }
+    cmd.cb.take().unwrap().call_box((resp,));
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: can we extract a function for `cmd.cb.take().unwrap().call_box` in PendingCmd, I see too many places use it.

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -1026,6 +1038,19 @@ impl Peer {
 
         Ok((resp, exec_result))
     }
+
+    /// Clear all the pending commands.
+    ///
+    /// Please note that all the pending callbacks will be lost.
+    /// Should not do this when dropping a peer in case of possible leak.
+    pub fn clear_pending_command(&mut self) {
+        while let Some(mut cmd) = self.pending_cmds.normals.pop_front() {
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: add log here if we still have pending callbacks

**Comment by BusyJay** [comment]
PTAL

**Comment by siddontang** [comment]
LGTM

**Comment by BusyJay** [comment]
@disksing PTAL

File: tests/raftstore/test_multi.rs [hunk 4]
```diff
@@ -369,15 +373,93 @@ fn test_leader_change_with_uncommitted_log<T: Simulator>(cluster: &mut Cluster<T
...
+                          assert!(resp.get_header().get_error().has_not_leader());
+                      })
+        .unwrap();
+
+    // Now let peer(1, 1) steps down. Can't use transfer leader here, because
+    // it still has pending proposed entries.
+    cluster.add_send_filter(CloneFilterFactory(RegionPacketFilter::new(1, 1)
+        .msg_type(MessageType::MsgHeartbeat)
+        .direction(Direction::Send)));
+    // peer 1 must have committed, but peer 2 has not.
```
Comment (disksing) [inline tests/raftstore/test_multi.rs]: I suppose the entry won't be committed on either 1 or 2.

**Comment by disksing** [comment]
LGTM

**Pull request merged**